### ISSUE TITLE
Improve error readibility

### ISSUE THREAD

**Issue by Atcold** [pytorch/pytorch#749]
The error PyTorch throws are a tad hard to parse.
For example I get the following
```
TypeError: FloatClassNLLCriterion_updateOutput received an invalid combination of arguments - got (int, torch.FloatTensor, torch.FloatTensor, torch.FloatTensor, bool, NoneType, torch.FloatTensor), but expected (int state, torch.FloatTensor input, torch.LongTensor target, torch.FloatTensor output, bool sizeAverage, [torch.FloatTensor weights or None], torch.FloatTensor total_weight)
```
It would be nicer to see something like "`target` is expected to be a `LongTensor`, got `FloatTensor` instead". Moreover, I couldn't figure out where the error has been thrown (which `assert` does this).

**Comment by soumith** [comment]
add to https://github.com/pytorch/pytorch/issues/39
